prov/util/src/util_mr_cache.c

```diff
@@ -0,0 +1,1493 @@
+			FI_WARN(UTIL_GET_PROV(cache), FI_LOG_MR,
+				"failed to erase lru entry from stale tree\n");
+			break;
+		}
+
+		util_mr_cache_entry_destroy(cache, entry);
+		entry = NULL;
+		++destroyed;
+	}
+
+	FI_DBG(UTIL_GET_PROV(cache), FI_LOG_MR,
+	       "flushed %i of %i entries from memory registration cache\n",
+	       destroyed, ofi_atomic_get32(&cache->stale.elements));
+
+	if (destroyed > 0)
+		ofi_atomic_sub32(&cache->stale.elements, destroyed);
+
+	return FI_SUCCESS;
+}
+
+int ofi_util_mr_cache_flush(struct util_mr_cache *cache)
+{
+	if (OFI_UNLIKELY(cache->state != UTIL_MRC_STATE_READY))
+		return -FI_EINVAL;
+
+	util_mr_cache_flush(cache, cache->attr.hard_reg_limit);
+
+	return FI_SUCCESS;
+}
+
+static int util_mr_cache_search_inuse(struct util_mr_cache *cache,
+				      uint64_t address,
+				      uint64_t length,
+				      struct util_mr_cache_entry **entry,
+				      struct util_mr_cache_key *key,
+				      struct util_fi_reg_context *fi_reg_context)
+{
+	int ret = FI_SUCCESS, cmp;
+	RbtIterator iter;
+	struct util_mr_cache_key *found_key, new_key;
+	struct util_mr_cache_entry *found_entry;
+	uint64_t new_end, found_end;
+	DEFINE_LIST(retired_entries);
+
+	if (cache->attr.lazy_deregistration)
+		util_mr_cache_clear_notifier_events(cache);
+
+	/* first we need to find an entry that overlaps with this one.
+	 * care should be taken to find the left most entry that overlaps
+	 * with this entry since the entry we are searching for might overlap
+	 * many entries and the tree may be left or right balanced
+	 * at the head
+	 */
+	iter = rbtFindLeftmost(cache->inuse.rb_tree, (void *) key,
+			util_find_overlapping_addr);
+	if (!iter) {
+		FI_DBG(UTIL_GET_PROV(cache), FI_LOG_MR,
+		       "could not find key in inuse, key=%"PRIu64":%"PRIu64"\n",
+		       key->address, key->length);
+		return -FI_ENOENT;
+	}
+
+	rbtKeyValue(cache->inuse.rb_tree, iter, (void **)&found_key,
+		    (void **)&found_entry);
+
+	FI_DBG(UTIL_GET_PROV(cache), FI_LOG_MR,
+	       "found a key that matches the search criteria, "
+	       "found=%"PRIu64":%"PRIu64" key=%"PRIu64":%"PRIu64"\n",
+	       found_key->address, found_key->length,
+	       key->address, key->length);
+
+	/* if the entry that we've found completely subsumes
+	 * the requested entry, just return a reference to
+	 * that existing registration
+	 */
+	if (util_can_subsume(found_key, key)) {
+		FI_DBG(UTIL_GET_PROV(cache), FI_LOG_MR,
+		       "found an entry that subsumes the request, "
+		       "existing=%"PRIu64":%"PRIu64" key=%"PRIu64":%"PRIu64
+		       "entry %p\n", found_key->address, found_key->length,
+		       key->address, key->length, found_entry);
+		*entry = found_entry;
+		util_mr_cache_entry_get(cache, found_entry);
+
+		cache->hits++;
+		return FI_SUCCESS;
+	}
+
+	/* otherwise, iterate from the existing entry until we can no longer
+	 * find an entry that overlaps with the new registration and remove
+	 * and retire each of those entries.
+	 */
```

Comment:
I've split this function into two ones:
- the first is responsible to search entry in inuse tree
- the second one looking for overlapped entries and if successful then adds the entry to the inuse tree.

Reply:
I'd really rather not push the feature back. The purpose of generalizing the code was so that all providers could use it. It would be a step backwards for the GNI provider (if we wanted to leverage common core -- and we support common core), as that functionality already exists and is tested. 

I agree that the process of growing a region should be moved into its own function though.